### DISCUSSION TITLE
Tailwind 적용 시 font-size 기반으로 rem 이 정해진다고 하는데,

### DISCUSSION THREAD

**Discussion by kimpearl3599** [GU-99/grow-up-fe#14] [질문과 답변]
font-size가 기본 16px 라고 가정했을 때,
예를 들어 h-4 라고 className을 작성하면 높이가 4rem인 요소가 생성이 되어야 하는데
1/4 인 1rem으로 생성이 되는 문제가 있어서

index.css 파일에 아래와 같이 기본 폰트 값을 추가해봤는데
해결이 안되네요...

![carbon (1)](https://github.com/GU-99/grow-up-fe/assets/80694130/dd0cdad0-1331-4b31-979f-431aa0b2a47c)

h-1 이면 높이가 1rem 
h-4 이면 높이가 4rem 이 되게 설정되게 하는 방법이 있을까요? @_@

**Comment by Yoonyesol** [comment]
진주님! tailwind 공식문서를 살펴보니 h-1이 0.25rem으로 4px이더라구요.  
![image](https://github.com/GU-99/grow-up-fe/assets/51500821/92c7f3a4-0047-411e-8d19-7ce004b007ac)
커스텀을 하려면 `tailwind.config.js` 파일에서 따로 수정을 좀 해줘야 합니다!
```
// tailwind.config.js
module.exports = {
  theme: {
    extend: {
      height: {
        '1': '1rem',  // 1rem (16px)
        '4': '4rem',  // 4rem (64px)
      },
    },
  },
  variants: {},
  plugins: [],
};

```

**Comment by Seok93** [comment]
html에 설정된 기본 font-size는 16px이고, 이 값이 1rem을 의미하잖아요?
tailwindcss에서 h-4는 1rem을 의미하니,  원하시는 4rem은 안나오리라 생각되어요.

만약에 h-4가 4rem이 되길 바란다면, 기존에 있는 속성을 오버라이딩 하셔야할 것 같아요.  아래와 같이 theme 속성에서 spacing 관련된 내용을 건드리면 할 수 있어요. ([customizing spacing 문서 참고](https://tailwindcss.com/docs/customizing-spacing))
```js
// tailwind.config.js
/** @type {import('tailwindcss').Config} */
export default {
  content: ['./index.html', './src/**/*.{js,ts,jsx,tsx}'],
  theme: {
    spacing: {
      1: '1rem',
      2: '2rem',
      3: '3rem',
      4: '4rem',
    },
  },
  plugins: [],
};
```

그런데 커스텀 관련된 문서를 보다보면 직접적으로 설정시, 기존의 내용이 사라진다고 하네요. 그래서 다른 내용을 유지하면서 원하는 부분을 추가하려면 기존의 내용을 상속하도록 만들어야 해요.

추가적으로 tailwid.config.js가 js 파일이고, export default로 반환하는 데이터가 객체잖아요? js 문법을 이용하여 특정 부분에 반복적은 속성 정의를 이용하는 방법도 있어요. ([comento tailwind: 우리 디자인에 맞게 커스텀하기 참조](https://developer.comento.kr/post/frontend-tailwind-css-24-04-01))

**Comment by kimpearl3599** [comment]
혹시 spacing 커스텀에 대해서는 예솔님, 우석님 의견이 어떤지 들어보고 싶습니다.
현 상태로 써도 큰 문제는 없지만, 직관성이 조금 떨어지는 느낌이라서요 :) 의견주세용!

**Reply by kimpearl3599** [reply]
spacing 을 기존보다 크게 설정하면
단위가 1rem 미만으로 작아지면 문제가 되겠죠..? 문득 그런 생각이 드네요

**Reply by Seok93** [reply]
진주님은 직관성을 어떻게 개선하고 싶으신건가요? 혹시 생각하시는 방법이 있나요?

음... tailwind 설정을 바꾸는 방법으로는 몇가지 방법이 있을 것 같아요.
1. rem을 기준으로 하기 때문에, rem의 기준이 되는 html의 font-size를 조절한다.
2. spacing 설정을 모두 덮어씌워서 새로운 크기 기준으로 만든다.
3. spacing 설정은 그대로 사용하고, 속성 확장으로 부족한 부분을 채워준다.
4. spacing 설정을 모두 없애고, 속성을 새롭게 정의한다.

진주님이 생각하시는 기준에 따라서 선택하면 좋을 것 같아요.
p.s 사실 저는 3번 방법으로 사용하려고 생각하고 있었어요

**Reply by kimpearl3599** [reply]
원래는 1번을 생각했는데, font-size를 조절하는 방법으로 적용이 어려워서
2번으로 설정하면 className에 작성하는 css랑 실제 단위가 똑같아서 작성하기나 보기에 편할 것 같다는 생각은 들어요!
3번은 색상이나 추가로 확장이 필요한 부분에 대해서는 적용하면 편리할 것 같습니다 :)

**Reply by Seok93** [reply]
아하 진주님이 생각하시는건, 붙어 있는 숫자가 곧 크기를 의미해줬으면 좋겠다는 말씀이시죠? 예를 들어서 `h-10` 이면1.0rem, `h-12`면 1.2rem 이런식으로요.

그러면 4번이 더 좋지 않을까요? spacing 기본 설정을 전부 없애고, 숫자를 소수점 한자리를 의미하도록 만들어서 연속적으로 만드는 방법을 사용하면 좋을 것 같아요

**Reply by Seok93** [reply]
음... 생각해보니 1/2이나 full은 유용하니 그대로 사용하는게 좋겠네요. 2번 3번을 적절히 섞어야할 것 같아요

**Reply by kimpearl3599** [reply]
`h-10` -> 10rem
`h-4` -> 4rem
이렇게 생각하긴 했는데, 단위가 너무 커서 문제가 된다면 우석님 말씀처럼 1/10로 줄이는 방법도 괜찮을 것 같습니다.

**Reply by Seok93** [reply]
음... 그럼 더 직관성을 높이는 방법으로 font-size를 10px로 설정하고, 단위를 1/10으로 낮춰서 사용해보는건 어떨까요?

```
h-10 = 1rem = 10px
h-25 = 2.5rem = 25px
```

**Reply by kimpearl3599** [reply]
font-size는 줄이지 말고, 
16px(font-size) = 1rem = h-10
이렇게 설정하는게 좋지 않을까 생각합니다.

기본 font-size를 변경하면 font도 커스텀을 해줘야할 것 같아서
더 혼란스러울 것 같아요...@_@

그래서 spacing만 건드리는게 좋지 않을까 싶습니다!!

**Reply by Yoonyesol** [reply]
좋습니다! 커스텀 과정의 불편함만 좀 감수한다면 직관성 면에서는 훨씬 코드 가독성이 좋아질 것 같아요👍

**Reply by Seok93** [reply]
@kimpearl3599 
그렇겠네요 ㅋㅋㅋㅋㅋ

root font-size를 10px로 설정하고, 단위를 1/10로 하려고 했다면, 처음부터 고려했어야 했네요 ㅎㅎ...

그럼 진주님 말씀대로 16px은 고정하고, spacing 단위만 커스텀 해볼게요.  0부터 40rem까지 가능하도록 한 번 설정해볼게요. (팝업 최소 크기를 600px로 고려하고 있고, 이 값은 37.5rem이니 편의상 40rem까지 만드려고 생각합니다.)

**Reply by kimpearl3599** [reply]
매번 사용할 때마다 *4를 해줘야 하는 불편함이 있는 것 같아서
생각해봤는데 제가 일을 키운 건 아닌지... 염려스럽네요 ㅠ
혹시 우석님 현재 만들고 있는 css 모듈들 샘플로 살짝 볼 수 있을까요??

**Reply by Seok93** [reply]
```js
// tailwind.config.js
/** @type {import('tailwindcss').Config} */
export default {
  content: ['./index.html', './src/**/*.{js,ts,jsx,tsx}'],
  theme: {
    extend: {
      spacing: customSpacing,
    },
  },
  plugins: [],
};

export const customSpacing = {};
for (let i = 0; i <= 400; i++) {
  customSpacing[i] = `${(i * 0.1).toFixed(1)}rem`;
}
```
위와 같이 변수로 만들어서 설정하려고 생각하고 있어요

**Reply by Seok93** [reply]
![커스텀 CSS 설정 내용](https://github.com/GU-99/grow-up-fe/assets/32609010/fa856d30-8527-4c79-9746-8e19a6291eb0)

다른 내용이랑 합치면 이런식으로 나올 것 같아요 (아직 설정 중이라서 예시로만 봐주세요!)

**Reply by kimpearl3599** [reply]
자료 찾아보니까 아이콘도 별도로 모아서 세팅하는 방식으로 사용하더라구요....
css 세계란... 엄청난 것 같습니다 수고 많으셨어요 >_<
